File: BloomFilterPython/MurmurHash.py
class MurmurHasher:    
    def hash_string(self,string_to_hash):
        c1 = 0xcc9e2d51
        c2 = 0x1b873593
        r1 = 15
        r2 = 13
        m = 5
        n = 0xe6546b64
        seed = 0
        lenght=len(string_to_hash)
        h = seed
        lenght4=lenght/4

        for i in range(0,lenght4):
          #little endian load order
            k=(ord(string_to_hash[i+3])<<24)^(ord(string_to_hash[i+2])<<16)^(ord(string_to_hash[i+1])<<8)^(ord(string_to_hash[i]))
            k *= c1
            k=k%4294967296#sve ovakve naredbe osiguravaju rad u artimetici broja 2^32
            k = (k << r1) | (k >> 32-r1)# ROTL32(k1,r1)
            k *= c2
            k=k%4294967296                                                                                                          
            h ^= k                                                                                                          
            h = (h << r2) | (h >> 32-r2)#ROTL32(h1,r2)
            h = h*5+0xe6546b64
            k=k%4294967296

        k=0

        if((lenght & 0x03)==3):
            k=ord(string_to_hash[lenght4 + 2])<<16

        if((lenght & 0x03)>=2):
            k=k^(ord(string_to_hash[lenght4 + 1])<<8)
        if((lenght & 0x03)>=1):
            k=k^(ord(string_to_hash[lenght4]))
            k *= c1
            k=k%4294967296
            k = (k<<r1) | (k>>32-r1)# ROTL32(k1,15);
            k *= c2
            k=k%4294967296
            h ^= k        

        #32bitna finalizacija
        h^= lenght
        h^= h >> 16
        h*= 0x85ebca6b
        h=h%4294967296
        h^= h>>13
        h*=0xc2b2ae35;
        h=h%4294967296
        h^= h >> 16

        return h%4294967296

File: BloomFilterPython/MurmurHash.java
package hr.fer.bioinformatika.projekt.hash;

public class MurmurHash implements HashFunction {
	protected static int c1 = 0xcc9e2d51;
	protected static int c2 = 0x1b873593;
	protected static int r1 = 15;
	protected static int r2 = 13;
	protected static int m = 5;
	protected static int n = 0xe6546b64;
	protected int seed = 0;

	@Override
	public int getHash(String s) {
		if (s == null) {
			throw new NullPointerException("Input string must not be null!");
		}
		byte[] data = null;
		try {
			data = s.getBytes();
		} catch (Exception e) {
			data = s.getBytes();
		}
		return calculate(data, 0, data.length, seed);
	}

	@Override
	public int additionalStep(int value) {
		// dodatno raspršivanje
		int hash = value;
		hash ^= 4;
		hash ^= hash >>> 16;
		hash *= 0x85ebca6b;
		hash ^= hash >>> 13;
		hash *= 0xc2b2ae35;
		hash ^= hash >>> 16;
		return hash;
	}

	public int calculate(byte[] data, int offset, int len, int seed) {
		int hash = seed;
		int roundedLen = offset + (len & 0xfffffffc);

		for (int i = offset; i < roundedLen; i += 4) {
			// little endian order
			int k = (data[i] & 0xff) | ((data[i + 1] & 0xff) << 8)
					| ((data[i + 2] & 0xff) << 16) | (data[i + 3] << 24);
			k *= c1;
			k = (k << r1) | (k >>> (32 - r1));
			k *= c2;
			hash ^= k;
			hash = (hash << r2) | (hash >>> r2);
			hash = hash * m + n;
		}

		// remaining bytes
		int k = 0;
		switch (len & 0x03) {
		case 3:
			k = (data[roundedLen + 2] & 0xff) << 16;
		case 2:
			k |= (data[roundedLen + 1] & 0xff) << 8;
		case 1:
			k |= (data[roundedLen] & 0xff);
		}
		k *= c1;
		k = (k << 15) | (k >>> 17);
		k *= c2;
		hash ^= k;

		// raspršivanje?
		hash ^= len;
		hash ^= hash >>> 16;
		hash *= 0x85ebca6b;
		hash ^= hash >>> 13;
		hash *= 0xc2b2ae35;
		hash ^= hash >>> 16;
		return hash & 0x7fffffff;
	}

}
